relay-server/src/actors/project_buffer.rs

```diff
@@ -321,32 +372,68 @@ impl BufferService {
-            )
-            .bind(key.own_key.to_string())
-            .bind(key.sampling_key.to_string())
-            .bind(100)
-            .fetch(db);
-
-                    while let Some(row) = envelopes.try_next().await? {
-                        let envelope_bytes_slice: &[u8] = row.try_get("envelope")?;
-                        let envelope_bytes = bytes::Bytes::from(envelope_bytes_slice);
-                        let envelope = Envelope::parse_bytes(envelope_bytes)?;
-                        let managed_envelope = ManagedEnvelope::new(
-                            envelope,
-                            permits.pop().ok_or(BufferError::Overloaded)?,
-                        );
-
-                        sender.send(managed_envelope).ok();
+                        "DELETE FROM envelopes WHERE id IN (SELECT id FROM envelopes WHERE own_key = ? AND sampling_key = ? LIMIT ?) RETURNING envelope",
+                    )
+                    .bind(key.own_key.to_string())
+                    .bind(key.sampling_key.to_string())
+                    .bind(request_size as u32)
+                    .fetch(db);
+
+                    loop {
+                        match envelopes.try_next().await {
+                            Ok(Some(row)) => match Self::extract_envelope(row, permits.pop()) {
+                                Ok(managed_envelope) => {
+                                    sender.send(managed_envelope).ok();
+                                }
+                                Err(err) => relay_log::error!(
+                                    "failed to extractt envelope from the buffer: {}",
+                                    LogError(&err)
+                                ),
+                            },
+                            Ok(None) => break,
+                            Err(err) => {
+                                relay_log::error!(
+                                    "failed to read the buffer stream from the disk: {}",
+                                    LogError(&err)
+                                );
+                                break;
+                            }
+                        }
+                    }
+
+                    // let's check if there are any data left in the db
+                    let result = sqlx::query(
+                        "SELECT id FROM envelopes WHERE own_key = ? AND sampling_key = ? LIMIT 1",
+                    )
+                    .bind(key.own_key.to_string())
+                    .bind(key.sampling_key.to_string())
+                    .fetch_one(db)
+                    .await;
+
+                    // Make sure to save the key which will be sent back if there are some more
+                    // records left in the db.
+                    if result.map_or(true, |row| !row.is_empty()) {
+                        back_keys.insert(key);
```

Comment:
I've rewritten this condition a little bit, to make cleat when exactly we have to hit DB again to check if there is more data for a key: 
1. when error occurs while fetching the data 
2. since we read in batches, which in my opinion we must do, otherwise we could lose a lot of data if something goes wrong, or we just push way to match data into the memory (and also we have to give fair share to all the keys which must be fetched from disk) - in this case we still have to check DB if there is data left and if there is, we should keep the key  around for the next round

Reply:
Thanks! I think I misread the nested loop on my last review. Instead of delegating `back_keys` to the project cache  in the `Ok(None)` case, why can't we keep requesting batches until there's nothing in the db? Something like

```
FOR EVERY key
    LOOP
        envelopes := fetch_and_delete_batch(batch_size=100)
        IF envelopes.is_empty()
            BREAK
        handle_envelopes(envelopes)
```

Reply:
@jjbayer  done in [6bc057b](https://github.com/getsentry/relay/pull/1920/commits/6bc057b6f66070b601529ae11b0cf6d615944972), please, take another look if that looks better for you.